Civi/ActionSchedule/RecipientBuilder.php

```diff
@@ -138,15 +138,15 @@ public function __construct($now, $actionSchedule, $mapping) {
   public function build() {
     $this->buildRelFirstPass();
 
-    if ($this->prepareAddlFilter('c.id') && $this->notTemplate()) {
+    if ($this->prepareAddlFilter('c.id') && $this->mapping->sendToAdditional($this->actionSchedule->entity_value)) {
```

Comment:
I started by reading 16311 and then this PR. I like how this one has a (REF) to delegate the check from the `RecipientBuilder` to each `ActionMapping`. 

In either revision, there's a couple things which feel odd/fishy. Just thinking out loud.

1. If a `civicrm_event` is generally not actionable because of `is_template=1`, and if you can make the short-circuit decision here, then shouldn't *all* processing be skipped (`buildRelFirstPass()`, `buildAddlFirstPass()`, `buildRelRepeatPass()`, `buildAddlRepeatPass()`)? It seems the value of`is_template` would be equally germane to all phases?
2.  In [the underlying createQuery()](https://github.com/MegaphoneJon/civicrm-core/blob/5b4d40abc013b45bf9ca3882210ba752f6a52420/CRM/Event/ActionMapping.php#L171-L172), it has a filter on `is_template` and `is_active`.  My expectation was that the filter would protect all four passes. (`createQuery()` is supposed to be a common building-block by way of `RecipientBuilder::build{Rel,Addl}{First,Repeat}}Pass()` => `RecipientBuilder::prepareQuery()` => `ActionMapping::createQuery()`). But evidently not...
3. I suppose there's a question about whether reminders should be sent for an inactive event. I would generally think not, and the `createQuery()` says no (`r.is_active = 1`)... so if we have to reproduce these guards, perhaps `sendToAdditional()` should also reproduce the `is_active` guard?

Anyway, these comments aren't meant as hard blocks. As a (REF) and a bugfix, the new code seems better than the old code. I'm just commenting because something about the context feels fishy.

Reply:
The heart of the issue here is that `createQuery()` isn't called when sending to additional recipients.  Which is why these passes are necessary to begin with - `createQuery()`'s docblock reads: `Generate a query to locate recipients who match the given schedule.`
Additional recipients are by definition arbitrary and outside the normal query parameters.

I hope that addresses your points 1 and 2!  I agree re: point 3, and I'll fix that when I clean up these failing tests.

Reply:
> The heart of the issue here is that createQuery() isn't called when sending to additional recipients.

Yeah, I had that intuition -- that maybe `createQuery()` isn't called because AddlFirst/AddlRepeat can use a static contact-list. But then I tried to check the intuition:

1. At least in skimming the PHP, there is a code-path from `buildAddl{First,Repeat}Pass()` to `prepareQuery()` to `createQuery()`  to `$query->where('r.is_template = 0');`. As near I can see, the path doesn't have any obstructions/short-circuits. It could just be my bad reading though. I guess a debugger would reveal the truth. Or maybe there's something in the structure of the query (like an `OR` or `IN` or `UNION`) which makes the filter inoperative.

2. AddlFirst/AddlRepeat do need *some* information from the `ActionMapping` - i.e. to identify the subject-matter for the reminder (the entity and the date; eg `civicrm_event`, `123`, `start_date`, `2020-06-02`). I suppose that some mapping-types (`EVENT_TPL_MAPPING_ID`, `EVENT_NAME_MAPPING_ID`) do convey part of the subject-matter (entity ID; e.g. `civicrm_event`, `123`) in their `$schedule` record. But `$schedule->entity_value` may not have entities. It could equally be an array of even-type-ids or  activity-type-ids  (if the mapping were `EVENT_TYPE_MAPPING_ID` or `ACTIVITY_MAPPING_ID`, respectively).

(IMHO, `entity_value` is a misnomer... it's more like "the primary filter supplied by the user" or `primary_query_value`.)

Tangentially... there may be a crossed wire in using `$schedule->entity_value` to set `sendToAdditional(int $entityId)`. Since `$schedule->entity_value` can have different things packed into it, it takes some work to unpack. Notice the extra work that `createQuery()` does [here](https://github.com/civicrm/civicrm-core/blob/4b99c9f3f207a6fd536080d63725f626a90b2942/CRM/Event/ActionMapping.php#L135) and [here](https://github.com/civicrm/civicrm-core/blob/4b99c9f3f207a6fd536080d63725f626a90b2942/CRM/Event/ActionMapping.php#L162-L169).

Reply:
I agree with @totten . How about simply passing the entityValue (which may be event ID(s) or event_type ID(s) based on ```EVENT_TPL_MAPPING_ID```) and extending the SQL here a bit 


 ```diff
diff --git a/CRM/Event/ActionMapping.php b/CRM/Event/ActionMapping.php
index 77fa8d77ed..9a91b69e07 100644
--- a/CRM/Event/ActionMapping.php
+++ b/CRM/Event/ActionMapping.php
@@ -186,4 +186,24 @@ class CRM_Event_ActionMapping extends \Civi\ActionSchedule\Mapping {
+  /**
+   * Determine whether a schedule based on this mapping should
+   * send to additional contacts.
+   */
+  public function sendToAdditional($entityValue): bool {
+    $selectedValues = (array) \CRM_Utils_Array::explodePadded($entityValue);
+    $valueField = ($this->id == \CRM_Event_ActionMapping::EVENT_TYPE_MAPPING_ID) ? 'event_type_id' : 'id';
+    // Don't send to additional recipients if this event is deleted or a template.
+    $query = new \CRM_Utils_SQL_Select('civicrm_event e');
+    $sql = $query
+      ->select('e.id')
+      ->where("e.{$valueField} IN (@selectedValues)")
+      ->where("is_template = 0")
+      ->where("is_active = 1")
+      ->param('selectedValues', $selectedValues)
+      ->toSQL();
+    $dao = \CRM_Core_DAO::executeQuery($sql);
+    return (bool) $dao->N;
+  }
+
 }
```
^^ @MegaphoneJon

Reply:
I went to implement this, then realized that the code Tim mentions doesn't account for "event template" scheduled reminders - just "event type" and "event name" scheduled reminders.  I implemented similar code, but it's a bit larger to support all three cases.

Reply:
@monishdeb can you review again the updated PR?